File: extras/vscode-pitchfork/src/dirs.ts
/**
 * Module to create directories
 */ /** */

import * as path from 'path';

import {BuildSystem, NewProjectParams} from './new';
import {fs} from './pr';

export function dirForNamespace(ns: string): string { return ns.replace(/::/g, '/'); }

export async function createDirectories(basePath: string, params: NewProjectParams) {
  const pfDir = path.join(basePath, params.name);
  // Create the directory where the project will live
  await fs.mkdir_p(pfDir);
  async function mkdir(subdir: string) { return fs.mkdir_p(path.join(pfDir, subdir)); }
  // Create the source directory
  const srcDir = path.join(pfDir, 'src');
  // And all the subdirectories based on the namespace
  await fs.mkdir_p(path.join(srcDir, dirForNamespace(params.rootNamespace)));
  // Generate the include/ dir, if applicable
  if (params.separateHeaders) {
    const incDir = path.join(pfDir, 'include');
    await fs.mkdir_p(path.join(incDir, dirForNamespace(params.rootNamespace)));
  }
  if (params.generateThirdParty) {
    await mkdir('third_party');
  }
  if (params.generateExamples) {
    await mkdir('examples');
  }
  await mkdir('tests');
  await mkdir('doc');
  if (params.buildSystem === BuildSystem.CMake) {
    await mkdir('cmake');
  }
}

File: src/pf/new/dirs.cpp
#include "./dirs.hpp"

#include <pf/new/project.hpp>

void pf::create_directories(const pf::new_project_params& params) {
    // Create the root directory
    fs::create_directories(params.directory);
    // Required subdirectories:
    fs::create_directories(params.directory / "src");
    fs::create_directories(params.directory / "docs");
    // Conditional subdirs:
    if (params.separate_headers) {
        fs::create_directories(params.directory / "include");
    }
    if (params.create_third_party) {
        fs::create_directories(params.directory / "third_party");
    }
    if (params.create_examples) {
        fs::create_directories(params.directory / "examples");
    }
    if (params.create_extras) {
        fs::create_directories(params.directory / "extras");
    }
    if (params.create_tests) {
        fs::create_directories(params.directory / "tests");
    }
    // Build system
    if (params.build_system == pf::build_system::cmake) {
        fs::create_directories(params.directory / "cmake");
    }
}
